api/desecapi/tests/test_token_domain_policy.py

```diff
@@ -0,0 +1,395 @@
+
+        ## Get
+        response = self.client.get_policy(self.token, using=self.token, domain=None)
+        self.assertStatus(response, status.HTTP_200_OK)
+        self.assertEqual(response.data, self.default_data | data)
+
+        # Inspection of other tokens forbitten
+        ## List
+        response = self.client.list_policies(self.token_manage, using=self.token)
+        self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+        ## Get
+        response = self.client.get_policy(self.token_manage, using=self.token, domain=None)
+        self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+        # Write operations forbidden (self and other)
+        for target in [self.token, self.token_manage]:
+            # Create
+            response = self.client.create_policy(target, using=self.token)
+            self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+            # Change
+            data = dict(domain=self.my_domains[1].name, perm_dyndns=False, perm_rrsets=True)
+            response = self.client.patch_policy(target, using=self.token, domain=self.my_domains[0].name, data=data)
+            self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+            # Delete
+            response = self.client.delete_policy(target, using=self.token, domain=None)
+            self.assertStatus(response, status.HTTP_403_FORBIDDEN)
+
+    def test_policy_lifecycle(self):
+        # Can't do anything unauthorized
+        response = self.client.list_policies(self.token, using=None)
+        self.assertStatus(response, status.HTTP_401_UNAUTHORIZED)
+
+        response = self.client.create_policy(self.token, using=None)
+        self.assertStatus(response, status.HTTP_401_UNAUTHORIZED)
+
+        # Create
+        ## without required field
+        response = self.client.create_policy(self.token, using=self.token_manage)
+        self.assertStatus(response, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data['domain'], ['This field is required.'])
+
+        ## without a default policy
+        data = dict(domain=self.my_domains[0].name)
+        with transaction.atomic():
+            response = self.client.create_policy(self.token, using=self.token_manage, data=data)
+        self.assertStatus(response, status.HTTP_400_BAD_REQUEST)
+        self.assertEqual(response.data['domain'], ['Policy precedence: The first policy must be the default policy.'])
+
+        # List: still empty
+        response = self.client.list_policies(self.token, using=self.token_manage)
+        self.assertStatus(response, status.HTTP_200_OK)
+        self.assertEqual(response.data, [])
+
+        # Create
+        ## default policy
+        data = dict(domain=None, perm_rrsets=True)
+        response = self.client.create_policy(self.token, using=self.token_manage, data=data)
+        self.assertStatus(response, status.HTTP_201_CREATED)
+
+        ## can't create another default policy
+        with transaction.atomic():
```

Comment:
why this is sometimes needed, sometimes not? Shouldn't atomicity be guaranteed by the API?

Reply:
from the code comment below, according to https://stackoverflow.com/a/23326971/6867099, this is only needed for status code >= 400, right?

Reply:
that is correct, and I believe it's what these tests are doing